torch/nn/modules/transformer.py

```diff
@@ -278,8 +278,21 @@ def forward(self, src: Tensor, mask: Optional[Tensor] = None, src_key_padding_ma
                 output = torch._nested_tensor_from_mask(output, src_key_padding_mask.logical_not(), mask_check=False)
                 src_key_padding_mask_for_layers = None
 
+        if mask is not None:
+            if _causal_mask:
+                raise RuntimeError("specify either mask or _causal_mask, but not both")
+
+            sz = mask.size(0)
+            causal_comparison = torch.triu(
+                torch.ones(sz, sz, device=mask.device) * float('-inf'), diagonal=1
+            ).to(mask.dtype)
```

Comment:
this might be very expensive, and will run every time the user provides a mask. It will also use quite a lot of memory, potentially negating the benefits from using memory-efficient kernels

Reply:
I think the preferred way will be to surface this as the causal mask flag as part of the API.  I am/was hoping that we might rely on torch.compile() to constant fold it.  (And maybe, if it is unable to do so, we might have it optimized away by torch.compile() ?